### PR TITLE
document exceptions for metrics/functional

### PR DESCRIPTION
**Pull request by dipam7**
## What does this PR do?

Part of #5434 

Files under metrics/functional:
- [x] __init__.py 
- [x] accuracy.py
- [x] auc.py
- [x] auroc.py
- [x] average_precision.py
- [x] classification.py
- [x] confusion_matrix.py
- [x] explained_variance.py
- [x] f_beta.py
- [x] hamming_distance.py
- [x] image_gradients.py
- [x] iou.py
- [x] mean_absolute_error.py
- [x] mean_relative_error.py
- [x] mean_squared_error.py
- [x] mean_squared_log_error.py
- [x] nlp.py
- [x] precision_recall.py
- [x] precision_recall_curve.py
- [x] psnr.py
- [x] r2score.py
- [x] roc.py
- [x] self_supervised.py
- [x] ssim.py
- [x] stat_scores.py

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/metrics/functional/precision_recall.py [hunk 1]
```diff
@@ -444,6 +468,18 @@ def precision_recall(
...
+        ValueError:
+            If ``average`` is not one of ``"micro"``, ``"macro"``, ``"weighted"``,
+            ``"samples"``, ``"none"`` or ``None``.
+        ValueError:
+            If ``mdmc_average`` is not one of ``None``, ``"samplewise"``, ``"global"``.
+        ValueError:
+            If ``average`` is set but ``num_classes`` is not provided.
+        ValueError:
+            If ``num_classes`` is set
+            and ``ignore_index`` is not in the range ``0`` <= ``ignore_index`` < ``num_classes``.
```
Comment (rohitgr7) [inline pytorch_lightning/metrics/functional/precision_recall.py:481]:
```suggestion
            and ``ignore_index`` is not in the range ``[0, num_classes)``.
```
can you change this everywhere?

File: pytorch_lightning/metrics/functional/psnr.py [hunk 2]
```diff
@@ -84,6 +84,10 @@ def psnr(
     Return:
         Tensor with PSNR score
 
+    Raises:
+        ValueError:
+            If ``dim`` is not None and ``data_range`` is not provided.
```
Comment (akihironitta) [inline pytorch_lightning/metrics/functional/psnr.py:89]:
```suggestion
            If ``dim`` is not ``None`` and ``data_range`` is not provided.
```

**Review by akihironitta — APPROVED** [review]
@dipam7 Once @rohitgr7's suggestion is resolved, LGTM!

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/metrics/functional/auroc.py [hunk 3]
```diff
@@ -165,6 +165,18 @@ def auroc(
             range [0, max_fpr]. Should be a float between 0 and 1.
         sample_weight: sample weights for each data point
 
+    Raises:
+        ValueError:
+            If ``max_fpr`` is not a ``float`` in the range ``0`` < max_fpr <= ``1``.
```
Comment (akihironitta) [inline pytorch_lightning/metrics/functional/auroc.py:170]:
```suggestion
            If ``max_fpr`` is not a ``float`` in the range ``(0, 1]``.
```

File: pytorch_lightning/metrics/functional/precision_recall.py [hunk 4]
```diff
@@ -133,6 +133,18 @@ def precision(
...
```
Comment (akihironitta) [inline pytorch_lightning/metrics/functional/precision_recall.py:146]:
```suggestion
            and ``ignore_index`` is not in the range ``[0, num_classes)``.
```

File: pytorch_lightning/metrics/functional/precision_recall.py [hunk 5]
```diff
@@ -295,6 +307,18 @@ def recall(
...
```
Comment (akihironitta) [inline pytorch_lightning/metrics/functional/precision_recall.py:320]:
```suggestion
            and ``ignore_index`` is not in the range ``[0, num_classes)``.
```

File: pytorch_lightning/metrics/functional/stat_scores.py [hunk 6]
```diff
@@ -244,6 +244,21 @@ def stat_scores(
...
+    Raises:
+        ValueError:
+            If ``reduce`` is none of ``"micro"``, ``"macro"`` or ``"samples"``.
+        ValueError:
+            If ``mdmc_reduce`` is none of ``None``, ``"samplewise"``, ``"global"``.
+        ValueError:
+            If ``reduce`` is set to ``"macro"`` and ``num_classes`` is not provided.
+        ValueError:
+            If ``num_classes`` is set
+            and ``ignore_index`` is not in the range ``0`` <= ``ignore_index`` < ``num_classes``.
```
Comment (akihironitta) [inline pytorch_lightning/metrics/functional/stat_scores.py:256]:
```suggestion
            and ``ignore_index`` is not in the range ``[0, num_classes)``.
```

File: pytorch_lightning/metrics/functional/precision_recall.py (hunk 1 above)
Comment (akihironitta) [reply pytorch_lightning/metrics/functional/precision_recall.py]: @dipam7 Applied this to other places for you.

**Review by akihironitta — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**